firedrake/preconditioners/hypre_ams.py

```diff
@@ -0,0 +1,77 @@
+from firedrake.petsc import PETSc
+from firedrake import FunctionSpace, Constant, project, Interpolator, grad, TestFunction
+from firedrake.assemble import allocate_matrix, create_assembly_callable
+
+__all__ = ("HypreAMS")
+
+class HypreAMS(firedrake.PCSNESBase):
+    def initialize(self, obj):
+        if isinstance(obj, PETSc.PC):
+            A, P = obj.getOperators()
+        elif isinstance(obj, PETSc.SNES):
+            A, P = obj.ksp.pc.getOperators()
+        else:
+            raise ValueError("Not a PC or SNES?")

```

Comment:
I think this should inherit from `PCBase` and not handle nonlinear things at all.